mapper/orm/src/main/java/org/hibernate/search/mapper/orm/batchindexing/impl/BatchIndexingWorkspace.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.mapper.orm.batchindexing.impl;
+
+import java.io.Serializable;
+import java.lang.invoke.MethodHandles;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadPoolExecutor;
+
+import org.hibernate.CacheMode;
+import org.hibernate.engine.spi.SessionFactoryImplementor;
+import org.hibernate.search.mapper.orm.logging.impl.Log;
+import org.hibernate.search.mapper.orm.mapping.spi.HibernateOrmMapping;
+import org.hibernate.search.mapper.orm.session.spi.HibernateOrmSearchManager;
+import org.hibernate.search.mapper.pojo.work.spi.PojoWorkPlan;
+import org.hibernate.search.util.AssertionFailure;
+import org.hibernate.search.util.impl.common.Executors;
+import org.hibernate.search.util.impl.common.LoggerFactory;
+
+/**
+ * This runnable will prepare a pipeline for batch indexing
+ * of entities, managing the lifecycle of several ThreadPools.
+ *
+ * @author Sanne Grinovero
+ */
+public class BatchIndexingWorkspace extends ErrorHandledRunnable {
+
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+	private final SessionFactoryImplementor sessionFactory;
+
+	private final ProducerConsumerQueue<List<Serializable>> primaryKeyStream;
+
+	private final int documentBuilderThreads;
+	private final Class<?> indexedType;
+	private final String idNameOfIndexedType;
+
+	// status control
+	private final CountDownLatch producerEndSignal; //released when we stop adding Documents to Index
+	private final CountDownLatch endAllSignal; //released when we release all locks and IndexWriter
+
+	// loading options
+	private final CacheMode cacheMode;
+	private final int objectLoadingBatchSize;
+
+	private final long objectsLimit;
+
+	private final int idFetchSize;
+	private final Integer transactionTimeout;
+
+	private final String tenantId;
+	private final PojoWorkPlan workPlan;
+
+	private final List<Future<?>> tasks = new ArrayList<>();
+
+	public BatchIndexingWorkspace(SessionFactoryImplementor sessionFactory,
+								HibernateOrmMapping mapping,
+								Class<?> type,
+								int objectLoadingThreads,
+								CacheMode cacheMode,
+								int objectLoadingBatchSize,
+								CountDownLatch endAllSignal,
+								long objectsLimit,
+								int idFetchSize,
+								Integer transactionTimeout,
+								String tenantId) {
+		this.indexedType = type;
+		this.idFetchSize = idFetchSize;
+		this.transactionTimeout = transactionTimeout;
+		this.tenantId = tenantId;
+
+		HibernateOrmSearchManager searchManager = mapping.createSearchManager( sessionFactory.createEntityManager() );
```

Comment:
This is awful, you're creating an entity manager but never closing it. You're effectively leaking JDBC connections here. Don't do that.

I don't see why you create a `WorkPlan` here. You should remove it. Here are the uses of this work plan I found:

* You pass it to `BatchTransactionalContext`, but it doesn't use it
* You pass it to `IdentifierConsumerDocumentProducer`, but that class is supposed to be thread safe, and the work plan is not. As we already discussed, the work plan in `IdentifierConsumerDocumentProducer` should be created in `org.hibernate.search.mapper.orm.batchindexing.impl.IdentifierConsumerDocumentProducer#run` (using the session you create in the same method) and passed as a parameter to other methods.